source/extensions/common/tap/admin.h

```diff
@@ -52,11 +63,122 @@ class AdminHandler : public Singleton::Instance,
   void unregisterConfig(ExtensionConfig& config);
 
   // Extensions::Common::Tap::Sink
-  PerTapSinkHandlePtr createPerTapSinkHandle(uint64_t) override {
-    return std::make_unique<AdminPerTapSinkHandle>(*this);
-  }
+  PerTapSinkHandlePtr
+  createPerTapSinkHandle(uint64_t trace_id,
+                         envoy::config::tap::v3::OutputSink::OutputSinkTypeCase type) override;
 
 private:
+  /**
+   * TraceBuffer internally buffers TraceWrappers in a vector. The size of the
+   * internal buffer is determined on construction by max_buf_size. TraceBuffer will continue to
+   * buffer traces via calls to bufferTrace until there is no remaining room in the buffer, or
+   * traceList is called, transfering ownership of the buffered data.
+   * Note that TraceBuffer is not threadsafe by itself - accesses to TraceBuffer need to be
+   * serialized. Serialization is currently done by the main thread dispatcher.
```

Comment:
consider adding calls to ASSERT_IS_MAIN_OR_TEST_THREAD() from source/common/common/thread.h if you want some asserts about that which disappear in product builds.